dynamodb/data.go

```diff
@@ -0,0 +1,161 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dynamodb
+
+import (
+	"encoding/json"
+	"fmt"
+	"math/big"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/dynamodb"
+	"github.com/cloudspannerecosystem/harbourbridge/internal"
+	"github.com/cloudspannerecosystem/harbourbridge/spanner/ddl"
+)
+
+// ProcessData performs data conversion for DynamoDB database. For each table,
+// we extract data using Scan requests, convert the data to Spanner data (based
+// on the source and Spanner schemas), and write it to Spanner. If we can't
+// get/process data for a table, we skip that table and process the remaining
+// tables.
+func ProcessData(conv *internal.Conv, client dynamoClient, consistentRead bool) error {
+	for srcTable, srcSchema := range conv.SrcSchema {
+		spTable, err1 := internal.GetSpannerTable(conv, srcTable)
+		spCols, err2 := internal.GetSpannerCols(conv, srcTable, srcSchema.ColNames)
+		spSchema, ok := conv.SpSchema[spTable]
+		if err1 != nil || err2 != nil || !ok {
+			conv.Stats.BadRows[srcTable] += conv.Stats.Rows[srcTable]
+			conv.Unexpected(fmt.Sprintf("Can't get cols and schemas for table %s: err1=%s, err2=%s, ok=%t",
+				srcTable, err1, err2, ok))
+			continue
+		}
+
+		var lastEvaluatedKey map[string]*dynamodb.AttributeValue
+
+		for {
+			// Build the query input parameters
+			params := &dynamodb.ScanInput{
+				TableName:      aws.String(srcTable),
+				ConsistentRead: &consistentRead,
+			}
+			if lastEvaluatedKey != nil {
+				params.ExclusiveStartKey = lastEvaluatedKey
+			}
+
+			// Make the DynamoDB Query API call
+			result, err := client.Scan(params)
+			if err != nil {
+				return fmt.Errorf("failed to make Query API call for table %v: %v", srcTable, err)
+			}
+
+			// Iterate the items returned
+			for _, attrsMap := range result.Items {
+				var cvtVals []interface{}
+				for i, srcColName := range srcSchema.ColNames {
+					var cvtVal interface{}
+					if attrsMap[srcColName] == nil {
+						cvtVal = nil
+					} else {
+						// Convert data to the target type
+						cvtVal, err = cvtColValue(attrsMap[srcColName], srcSchema.ColDefs[srcColName].Type.Name, spSchema.ColDefs[spCols[i]].T.Name)
+						if err != nil {
+							return fmt.Errorf("failed to convert column: %v to %v", attrsMap[srcColName], spSchema.ColDefs[spCols[i]])
```

Comment:
Do we want to return here?
That would mean we'd stop processing data on the first error we encountered.
Suggest add row to bad row stats and sampleBadRows and continuing.